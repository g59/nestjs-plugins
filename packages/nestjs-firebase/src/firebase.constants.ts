// biome-ignore lint/complexity/noStaticOnlyClass: <explanation>
export class FirebaseConstants {
  public static readonly FIREBASE_MODULE = "FirebaseModule";
  public static readonly FIREBASE_TOKEN = "FIREBASE_TOKEN";
}
