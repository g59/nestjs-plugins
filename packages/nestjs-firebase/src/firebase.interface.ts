import type { Type } from "@nestjs/common";
import type { ModuleMetadata } from "@nestjs/common/interfaces";
import * as firebaseAdmin from "firebase-admin";
import { AppOptions } from "firebase-admin";

export type FirebaseModuleOptions = {
  googleApplicationCredential?: string | firebaseAdmin.ServiceAccount;
} & Omit<AppOptions, "credential">;

export type FirebaseModuleAsyncOptions = {
  useClass?: Type<FirebaseModuleOptionsFactory>;
  useFactory?: (
    ...args: any[]
  ) => Promise<FirebaseModuleOptions> | FirebaseModuleOptions;
  inject?: any[];
  useExisting?: Type<FirebaseModuleOptionsFactory>;
} & Pick<ModuleMetadata, "imports">;

export interface FirebaseModuleOptionsFactory {
  createFirebaseModuleOptions():
    | Promise<FirebaseModuleOptions>
    | FirebaseModuleOptions;
}

export interface FirebaseAdmin {
  auth: firebaseAdmin.auth.Auth;
  messaging: firebaseAdmin.messaging.Messaging;
  firestore: firebaseAdmin.firestore.Firestore;
  storage: firebaseAdmin.storage.Storage;
  remoteConfig: firebaseAdmin.remoteConfig.RemoteConfig;
}
