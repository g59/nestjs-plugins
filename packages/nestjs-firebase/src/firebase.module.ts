import { DynamicModule, Global, Module, Provider } from "@nestjs/common";
import { ClassProvider } from "@nestjs/common/interfaces";
import { FirebaseConstants } from "./firebase.constants";
import {
  FirebaseAdmin,
  FirebaseModuleAsyncOptions,
  FirebaseModuleOptions,
  FirebaseModuleOptionsFactory,
} from "./firebase.interface";
import { getFirebaseAdmin } from "./util";

@Global()
@Module({})
// biome-ignore lint/complexity/noStaticOnlyClass: <explanation>
export class FirebaseModule {
  public static forRoot(options: FirebaseModuleOptions): DynamicModule {
    const provider: Provider<FirebaseAdmin> = {
      provide: FirebaseConstants.FIREBASE_TOKEN,
      useValue: getFirebaseAdmin(options),
    };

    return {
      exports: [provider],
      module: FirebaseModule,
      providers: [provider],
    };
  }

  public static forRootAsync(
    options: FirebaseModuleAsyncOptions,
  ): DynamicModule {
    const firebaseProvider: Provider = {
      inject: [FirebaseConstants.FIREBASE_MODULE],
      provide: FirebaseConstants.FIREBASE_TOKEN,
      useFactory: (options: FirebaseModuleOptions) => getFirebaseAdmin(options),
    };

    const asyncProviders = FirebaseModule.createAsyncProviders(options);
    return {
      module: FirebaseModule,
      imports: [...(options.imports || [])],
      providers: [...asyncProviders, firebaseProvider],
      exports: [firebaseProvider],
    };
  }

  private static createAsyncProviders(
    options: FirebaseModuleAsyncOptions,
  ): Provider[] {
    if (options.useFactory || options.useExisting) {
      return [FirebaseModule.createAsyncOptionsProvider(options)];
    }
    return [
      FirebaseModule.createAsyncOptionsProvider(options),
      {
        provide: options.useClass,
        useClass: options.useClass,
        inject: options.inject,
      } as ClassProvider,
    ];
  }

  private static createAsyncOptionsProvider(
    options: FirebaseModuleAsyncOptions,
  ): Provider {
    if (options.useFactory) {
      return {
        provide: FirebaseConstants.FIREBASE_MODULE,
        useFactory: options.useFactory,
        inject: options.inject || [],
      };
    }
    return {
      provide: FirebaseConstants.FIREBASE_MODULE,
      useFactory: async (
        optionsFactory: FirebaseModuleOptionsFactory,
      ): Promise<FirebaseModuleOptions> =>
        await optionsFactory.createFirebaseModuleOptions(),
      inject: options.useClass ? [options.useClass] : [],
    };
  }
}
