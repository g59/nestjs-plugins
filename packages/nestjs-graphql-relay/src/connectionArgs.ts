import { ArgsType, Field, Int } from "@nestjs/graphql";
import {
  Min,
  Validate,
  ValidateIf,
  ValidationArguments,
  ValidatorConstraint,
  ValidatorConstraintInterface,
} from "class-validator";
import * as Relay from "graphql-relay";
import "reflect-metadata";

@ValidatorConstraint({ async: false })
export class CannotUseWithout implements ValidatorConstraintInterface {
  validate(_: unknown, args: ValidationArguments) {
    const object = args.object as Record<string, unknown>;
    const required = args.constraints[0] as string;
    return object[required] !== undefined;
  }

  defaultMessage(args: ValidationArguments) {
    return `Cannot be used without \`${args.constraints[0]}\`.`;
  }
}

@ValidatorConstraint({ async: false })
export class CannotUseWith implements ValidatorConstraintInterface {
  validate(_: unknown, args: ValidationArguments) {
    const object = args.object as Record<string, unknown>;
    const result = args.constraints.every((propertyName) => {
      return object[propertyName] === undefined;
    });
    return result;
  }

  defaultMessage(args: ValidationArguments) {
    return `Cannot be used with \`${args.constraints.join("` , `")}\`.`;
  }
}

@ArgsType()
export class ConnectionArgs implements Relay.ConnectionArguments {
  @Field(() => String, {
    nullable: true,
    description: "Paginate before opaque cursor",
  })
  @ValidateIf((o) => o.before !== undefined)
  @Validate(CannotUseWithout, ["last"])
  @Validate(CannotUseWith, ["after", "first"])
  readonly before?: Relay.ConnectionCursor;

  @Field(() => String, {
    nullable: true,
    description: "Paginate after opaque cursor",
  })
  @ValidateIf((o) => o.after !== undefined)
  @Validate(CannotUseWithout, ["first"])
  @Validate(CannotUseWith, ["before", "last"])
  readonly after?: Relay.ConnectionCursor;

  @Field(() => Int, { nullable: true, description: "Paginate first" })
  @ValidateIf((o) => o.first !== undefined)
  @Min(1)
  @Validate(CannotUseWith, ["before", "last"])
  readonly first?: number;

  @Field(() => Int, { nullable: true, description: "Paginate last" })
  @ValidateIf((o) => o.last !== undefined)
  // Required `before`. This is a weird corner case.
  // We'd have to invert the ordering of query to get the last few items then re-invert it when emitting the results.
  // We'll just ignore it for now.
  @Validate(CannotUseWithout, ["before"])
  @Validate(CannotUseWith, ["after", "first"])
  @Min(1)
  readonly last?: number;
}
