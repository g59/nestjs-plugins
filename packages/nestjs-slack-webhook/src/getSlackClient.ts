import { IncomingWebhook } from "@slack/webhook";
import type { SlackOptions } from "./slackOptions";

export function getSlackClient({
  url,
  ...args
}: SlackOptions): IncomingWebhook {
  const slackClient = new IncomingWebhook(url, args);
  return slackClient;
}
