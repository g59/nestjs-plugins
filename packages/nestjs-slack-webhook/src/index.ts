export * from "./injectSlack";
export * from "./slackConstants";
export * from "./slackModule";
export * from "./slackOptions";
