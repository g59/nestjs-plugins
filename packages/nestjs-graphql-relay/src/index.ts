import * as Relay from "graphql-relay";
import { FindManyOptions, ObjectLiteral, Repository } from "typeorm";
import { ConnectionArgs } from "./connectionArgs";

export * from "./orderBy.input";
export * from "./pageInfo";
export { ConnectionArgs };

type PagingMeta =
  | { pagingType: "forward"; after?: string; first: number }
  | { pagingType: "backward"; before?: string; last: number }
  | { pagingType: "none" };

function getMeta({
  first = 0,
  last = 0,
  after,
  before,
}: ConnectionArgs): PagingMeta {
  const isForwardPaging = !!first || !!after;
  const isBackwardPaging = !!last || !!before;

  return isForwardPaging
    ? { pagingType: "forward", after, first }
    : isBackwardPaging
      ? { pagingType: "backward", before, last }
      : { pagingType: "none" };
}

/*
 * Create a 'paging parameters' object with 'limit' and 'offset' fields based on the incoming
 * cursor-paging arguments.
 */
export function getPagingParameters(args: ConnectionArgs) {
  const meta = getMeta(args);

  switch (meta.pagingType) {
    case "forward": {
      let offset = 0;
      if (meta.after) {
        offset = Relay.cursorToOffset(meta.after) + 1;
      }

      if (Number.isNaN(offset)) {
        throw new Error("invalid before query");
      }

      return {
        limit: meta.first,
        offset,
      };
    }
    case "backward": {
      const { last, before } = meta;
      let limit = last;
      let offset = Relay.cursorToOffset(before || "") - last;

      if (Number.isNaN(offset)) {
        throw new Error("invalid before query");
      }

      // Check to see if our before-page is underflowing past the 0th item
      if (offset < 0) {
        // Adjust the limit with the underflow value
        limit = Math.max(last + offset, 0);
        offset = 0;
      }

      return { offset, limit };
    }
    default:
      return {};
  }
}

export async function findAndPaginate<T extends ObjectLiteral>(
  condition: FindManyOptions<T>,
  connArgs: ConnectionArgs,
  repository: Repository<T>,
) {
  const { limit, offset } = getPagingParameters(connArgs);
  const [entities, count] = await repository.findAndCount({
    ...condition,
    skip: offset,
    take: limit,
  });

  return Relay.connectionFromArraySlice(entities, connArgs, {
    arrayLength: count,
    sliceStart: offset || 0,
  });
}
