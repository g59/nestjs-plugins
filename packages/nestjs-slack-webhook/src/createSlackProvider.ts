import type { Provider } from "@nestjs/common";
import type { IncomingWebhook } from "@slack/webhook";
import { getSlackClient } from "./getSlackClient";
import { SLACK_TOKEN } from "./slackConstants";
import type { SlackOptions } from "./slackOptions";

export function createSlackProvider(
  options: SlackOptions,
): Provider<IncomingWebhook> {
  return {
    provide: SLACK_TOKEN,
    useValue: getSlackClient(options),
  };
}
