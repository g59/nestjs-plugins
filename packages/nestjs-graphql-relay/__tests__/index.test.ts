import { random } from "faker";
import { Factory } from "typeorm-factory";
import {
  Entity,
  PrimaryGeneratedColumn,
  Column,
  getRepository,
  Connection,
  createConnection,
} from "typeorm";
import { getPagingParameters, findAndPaginate } from "../src";

@Entity()
class Example {
  @PrimaryGeneratedColumn()
  readonly id: number;

  @Column()
  readonly name: string;
}

describe("app", () => {
  let connection: Connection;
  beforeAll(async () => {
    connection = await createConnection({
      type: "sqlite",
      database: "nestjs-plugins",
      synchronize: true,
      entities: [Example],
    });

    const repo = getRepository(Example);
    await connection.query(`DELETE from ${repo.metadata.tableName}`);
  });

  afterAll(() => connection.close());

  it("getPagingParameters", () => {
    expect(getPagingParameters({})).toMatchInlineSnapshot(`Object {}`);
    expect(getPagingParameters({ first: 1 })).toMatchInlineSnapshot(`
      Object {
        "limit": 1,
        "offset": 0,
      }
    `);
    expect(() =>
      getPagingParameters({ first: 1, after: "after" })
    ).toThrowErrorMatchingInlineSnapshot(`"invalid before query"`);
    expect(() =>
      getPagingParameters({ last: 1, before: "before" })
    ).toThrowErrorMatchingInlineSnapshot(`"invalid before query"`);
  });

  describe("findAndPaginate", () => {
    it("empty", async () => {
      const res = await findAndPaginate(
        {
          where: { name: "undefined" },
        },
        {},
        getRepository(Example)
      );
      expect(res).toMatchInlineSnapshot(`
        Object {
          "edges": Array [],
          "pageInfo": Object {
            "endCursor": null,
            "hasNextPage": false,
            "hasPreviousPage": false,
            "startCursor": null,
          },
        }
      `);
    });

    it("find", async () => {
      const name = random.words();
      const f = new Factory(Example).attr("name", random.alphaNumeric(10));
      await f.createList(random.number({ max: 5 }));
      await f.createList(random.number({ max: 5 }), {
        name,
      });

      const res = await findAndPaginate(
        {
          where: { name },
        },
        {},
        getRepository(Example)
      );
      res.edges.map(({ node }) => expect(node.name).toEqual(name));
      expect(res.pageInfo).toMatchSnapshot();
    });
  });
});
