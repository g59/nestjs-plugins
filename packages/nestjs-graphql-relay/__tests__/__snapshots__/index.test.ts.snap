// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app findAndPaginate find 1`] = `
{
  "endCursor": "YXJyYXljb25uZWN0aW9uOjA=",
  "hasNextPage": false,
  "hasPreviousPage": false,
  "startCursor": "YXJyYXljb25uZWN0aW9uOjA=",
}
`;
